### PR TITLE
adding libpq/12.2

### LINKED ISSUE #1574: [request] libpq/12.2
### Package Details
  * Package Name/Version: **libpq/12.2**
  * Changelog: **https://github.com/postgres/postgres/releases**

The above mentioned version is newly released by the upstream project and not yet available as a recipe. Please add this version.

### PR DESCRIPTION
**Pull request by prince-chrismc**
fixes #1574

Specify library name and version:  **libpq/12.2**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
Not very familiar with autotools. How does one resolve build order?

The build failed because:
```
/usr/bin/ld: cannot find -lpgport_shlib
collect2: error: ld returned 1 exit status
../../../src/Makefile.shlib:309: recipe for target 'libpq.so.5.12' failed
make: *** [libpq.so.5.12] Error 1
make: *** Waiting for unfinished jobs....
```
Which included:
```
rm -f libpgport_shlib.a
ar crs libpgport_shlib.a fls_shlib.o strlcat_shlib.o strlcpy_shlib.o getpeereid_shlib.o pg_crc32c_sse42_shlib.o pg_crc32c_sb8_shlib.o pg_crc32c_sse42_choose_shlib.o chklocale_shlib.o erand48_shlib.o inet_net_ntop_shlib.o noblock_shlib.o path_shlib.o pg_bitutils_shlib.o pgcheckdir_shlib.o pgmkdirp_shlib.o pgsleep_shlib.o pg_strong_random_shlib.o pgstrcasecmp_shlib.o pgstrsignal_shlib.o pqsignal_shlib.o qsort_shlib.o qsort_arg_shlib.o quotes_shlib.o snprintf_shlib.o sprompt_shlib.o strerror_shlib.o tar_shlib.o thread_shlib.o
```

Both are apart of 
```py
            with tools.chdir(os.path.join(self._source_subfolder, "src", "interfaces", "libpq")):
                autotools.make()
```

Was I just unlucky?

**Comment by madebr** [comment]
@prince-chrismc
By calling the different sub Makefiles seperately, we probably broke a dependency because the Makefile scripts assumed that its dependencies were already built.
I can fix this on my pc by doing the following 
```python
            autotools = self._configure_autotools()
            with tools.chdir(os.path.join(self._source_subfolder, "src", "backend")):
                autotools.make(target="generated-headers")
            with tools.chdir(os.path.join(self._source_subfolder, "src", "common")):
                autotools.make()
            with tools.chdir(os.path.join(self._source_subfolder, "src", "port")):  # NEW LINE
                autotools.make()                                                    # NEW LINE
            with tools.chdir(os.path.join(self._source_subfolder, "src", "include")):
                autotools.make()
            with tools.chdir(os.path.join(self._source_subfolder, "src", "interfaces", "libpq")):
                autotools.make()
            with tools.chdir(os.path.join(self._source_subfolder, "src", "bin", "pg_config")):
                autotools.make()
```
(running `make` in `src/port`)

Also, I saw that the resulting package contains both static and shared libraries.
We probably need to filter out the static and shared libraries.
This is also the case with libpq/11.5.
Because of the interdependencies, we probably need to drop the `tools.collect_libs(self)`.
Also, the number of static libraries is different between 11.5 and 12.2...

**Comment by syoliver** [comment]
@madebr @prince-chrismc I made a proposal to merge #1633 to this PR here : [#3](https://github.com/prince-chrismc/conan-center-index/pull/3)

**Comment by prince-chrismc** [comment]
Thank you very much for helping. Nasty production bug killing me slowly

**Comment by prince-chrismc** [comment]
```
ERROR: Unserializable object <conans.model.build_info.Component object at 0x7f11b4edc510> of type <class 'conans.model.build_info.Component'>
```

Seems like the usage of components may be an issue in the pipeline

**Comment by prince-chrismc** [comment]
Might be https://github.com/conan-io/conan/pull/6966 which was released ~24hrs ago.

**Comment by madebr** [comment]
I thought 1.25.1 was already available on ci because my Poco pr uses components as well and succeeded.
@danimtb @uilianries Is it possible that some docker instances on cci are running an older version of Conan?

**Comment by madebr** [comment]
Indeed, same as https://github.com/conan-io/conan-center-index/pull/1541#issuecomment-624739446.
Let's restart this in a day or two.

**Comment by uilianries** [comment]
Unfortunately, the Conan version upgrade is not executed automatically, that's why sometimes you see errors from early versions.

**Comment by danimtb** [comment]
You might be interested on this: https://github.com/conan-io/conan-center-index/wiki/Deployment-Changelog#14-may-2020---1552-cest

Unfortunately, there is still no sync between Linux containers and the Linux and Mac agents but we are working on it! 😄

File: recipes/libpq/all/conanfile.py [hunk 1]
```diff
@@ -113,19 +114,33 @@ def build(self):
...
                 autotools.make()
 
+    def _remove_unused_libraries_from_package(self, is_shared):
+        path = None
+        if is_shared:
+            path = os.path.join(self.package_folder, "lib", "*.a")
+        else:
+            path = os.path.join(self.package_folder, "lib", "libpq.so*")
+        for file in glob.glob(path):
+            os.remove(file)
```
Comment (SpaceIm) [inline recipes/libpq/all/conanfile.py:142]: Does it properly clean up unused libraries with MinGW?

File: recipes/libpq/all/conanfile.py [hunk 2]
```diff
@@ -156,12 +179,44 @@ def package(self):
...
+                    self.cpp_info.components["pgport"].libs = ["pgport", "pgport_shlib"]
+                    self.cpp_info.components["pq"].requires.extend(["pgport", "pgcommon"])
+
+
         if self.settings.os == "Linux":
-            self.cpp_info.system_libs = ["pthread"]
+            self.cpp_info.components["pq"].system_libs = ["pthread"]
         elif self.settings.os == "Windows":
-            self.cpp_info.system_libs = ["ws2_32", "secur32", "advapi32", "shell32", "crypt32", "wldap32"]
+            self.cpp_info.components["pq"].system_libs = ["ws2_32", "secur32", "advapi32", "shell32", "crypt32", "wldap32"]
```
Comment (SpaceIm) [inline recipes/libpq/all/conanfile.py:222]: What is the cmake name of these components?

Comment (syoliver) [reply recipes/libpq/all/conanfile.py:222]: * PostgreSQL::pq : The library itself
* PostgreSQL::pgcommon
* PostgreSQL::pgport

File: recipes/libpq/all/conanfile.py (hunk 1 above)
Comment (syoliver) [reply recipes/libpq/all/conanfile.py:142]: Certainely not :+1:

Comment (prince-chrismc) [reply recipes/libpq/all/conanfile.py:142]: What file extension would need to be added?

Comment (prince-chrismc) [reply recipes/libpq/all/conanfile.py:142]: @syoliver would be so kind as to explain the solution? I have little to no experience with MinGW so if there's something wrong I would not be able to know.

Comment (syoliver) [reply recipes/libpq/all/conanfile.py:142]: I would like to propose something but I can't test it right now. I am not sure but I think that naming is :
- *.a : static library
- *.dll : dynamic library
- *.dll.a : import library

If you can't fix it, I will try to propose something... But not sure when.

Comment (prince-chrismc) [reply recipes/libpq/all/conanfile.py:142]: MinGW is not a supported platform, so it should not block this PR =)

Comment (syoliver) [reply recipes/libpq/all/conanfile.py:142]: @SpaceIm @madebr I don't know what to do with the following error, I did something bad ?

```
C:\Dev\conan-center-index\recipes\libpq\all>conan create . libpq/12.2@ --profile
=mingw --build
Exporting package recipe
libpq/12.2 exports: File 'conandata.yml' found. Exporting it...
libpq/12.2 exports: Copied 1 '.yml' file: conandata.yml
libpq/12.2: The stored package has not changed
libpq/12.2: Exported revision: 0c097720baf42d6d959a7dda17217f33
Configuration:
[settings]
arch=x86
arch_build=x86
build_type=Debug
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=9.2
os=Windows
os_build=Windows
[options]
[build_requires]
[env]

zlib/1.2.11: Forced build from source
libpq/12.2: Forced build from source
msys2/20190524: Forced build from source
7zip/19.00: Forced build from source
lzma_sdk/9.20: Forced build from source
make/4.2.1: Forced build from source
ERROR: Loop detected in context host: 'lzma_sdk/9.20' requires 'msys2/20190524'
which is an ancestor too
```

Comment (madebr) [reply recipes/libpq/all/conanfile.py:142]: @syoliver See https://github.com/conan-io/conan-center-index/issues/1422#issuecomment-637639426
Does removing the `msys2` build requirement for `lzma_sdk` fix this?

Comment (SpaceIm) [reply recipes/libpq/all/conanfile.py:142]: related to https://github.com/conan-io/conan-center-index/issues/1422

use `-b missing` instead of `--build`

Comment (syoliver) [reply recipes/libpq/all/conanfile.py:142]: Actually I think msys2 should not depend on 7zip. Here is the PR #1810

Comment (madebr) [reply recipes/libpq/all/conanfile.py:142]: @prince-chrismc 
Please consider https://github.com/prince-chrismc/conan-center-index/pull/4

**Comment by danimtb** [comment]
@prince-chrismc nice work there with components! Out of curiosity, do those components have a dedicated target in cmake? Is there any documentation of the library where I can check them?

I think it will be interesting to try to mimic the CMake target names in that case as well

**Comment by prince-chrismc** [comment]
I haven't a clue! I thought this was an easy bump version!

@syoliver and @madebr did all the heavy lifting :heart: 

From the quick search I did reviewing their changes they do appear to be the correct targets from the original build script =) once 1.26 is out this would be a great test for the new components feature.

The only caveat is that the [c make module](https://cmake.org/cmake/help/latest/module/FindPostgreSQL.html) from kitware does not expose individual targets.

**Comment by tomlankhorst** [comment]
Is there a way to test this recipe locally already?

**Comment by prince-chrismc** [comment]
> Is there a way to test this recipe locally already?

Of Course!

If you clone the repo, cd to `recipes/libpq/all` run `conan create . 1.12.0@` you will have a local copy!

**Comment by prince-chrismc** [comment]
> once 1.26 is out this would be a great test for the new components feature

Took the time to test this, with Conan 1.26, you can use the official CMake module imported target :+1:

**Pull request merged**